### ISSUE TITLE
Feature request/Idea: a way of not missing old notifications

### ISSUE THREAD

**Issue by tsujan** [lxqt/lxqt#1573]
Needless to say, notifications are temporary and can be missed. It would be a good feature if LXQt could save and show older notifications and have a setting for that, so that the user decides how many old notifications should be saved/shown.

Older notifications may be shown inside a scroll-view. The panel may need a new widget for showing them.

Is this acceptable to you?

**Comment by agaida** [comment]
Never missed that feature, but yes - sounds nice for people needing this

**Comment by Vladimir-csp** [comment]
Sometimes new ideas are just thoroughly forgotten old ideas: #149, #767.

**Comment by tsujan** [comment]
> Never missed that feature

I do, sometimes. Every 10 minutes, I get gmail notifications through a bash script and if it weren't for Conky (which shows new mails on desktop), I'd miss them when coming back to the computer after more than 10 minutes.

> Sometimes new ideas are just thoroughly forgotten old ideas: #149, #767.

Well said! I think this one is more general and can supersede them.

**Comment by agaida** [comment]
if there are any ideas in both bugs worth considering here - please add an excerpt and close the old bugs. Far more important - one should start and implement, talking about needed code helps noone

**Comment by agaida** [comment]
@tsujan - now that you showing your use case - yes, it is useful - and i would like it too.

**Comment by Vladimir-csp** [comment]
Also, #321 is related to temporal readability of notifications (timeouts for urgency levels).

**Comment by tsujan** [comment]
> if there are any ideas in both bugs worth considering here

Yes, they are:

(1) "A notification would go in there when it automatically fades out" (quoted from https://github.com/lxqt/lxqt/issues/149). I don't agree with this for obvious reasons.

(2) "Please add keyboard shortcuts to lxqt-notificaiond to close or bring back messages (one by one)." (quoted from https://github.com/lxqt/lxqt/issues/767). Not the best idea.

**Comment by tsujan** [comment]
> I bet you'd change your mind after using Dunst.

I will use it to get your point.

**Comment by Vladimir-csp** [comment]
More "GUI-friendly" way would be a tray icon, what else? And shortcut configuration/info could be easily available via right-click menu of the tray icon.

**Comment by tsujan** [comment]
> talking about needed code helps noone

@agaida, I agree. Will try to dedicate some time to the code.

Talking is good for knowing what's the best approach. Differences in view points can be productive. But if no one does anything (like in the case of font problem), talking won't make a miracle.

**Comment by tsujan** [comment]
> More "GUI-friendly" way would be a tray icon,

I didn't know it goes to the tray ;) As I said, I'll try Dunst.

**Comment by agaida** [comment]
And i want to have the Captain-Obvious-Award for the sentence above. Like this one:

* https://i.imgur.com/vsdxJo6.jpg

**Comment by tsujan** [comment]
The tray reminds me of https://github.com/lxqt/lxqt/issues/1012#issuecomment-366527477. I lost my work on it :( Too many jobs and a limited time...

**Comment by tsujan** [comment]
@agaida You should have meant this sentence: https://github.com/lxqt/qterminal/issues/488#issuecomment-429127753 :D

**Comment by Vladimir-csp** [comment]
1. Good idea.
2. In everyday usage many notifications reach timeout without being explicitly closed, so that distinction isn't as useful as as it seems. Just remember all of the last N notifications. Maybe gray out a bit those which were closed explicitly.
3. `$TIME ago` format may be useful here.

BTW, another feature of Dunst is inactivity timeout: notifications don't expire if user is AFK more that some number of seconds.

**Comment by tsujan** [comment]
@Vladimir-csp, thanks for the feedback!

> In everyday usage many notifications reach timeout without being explicitly closed

Yes, you're right but those explicitly closed are seen and read. That was @jleclanche's idea, which I misunderstood at first but now, I agree with him. In this way, there might be a little more chance to save other notifications. For example, my email notification has a button for launching Thunderbird and I often click that button.

> BTW, another feature of Dunst is inactivity timeout: notifications don't expire if user is AFK more that some number of seconds.

I like that feature. It can be the title of another feature request.

This might not be ready soon because I'm busy with other projects but I started to work on it.

**Comment by lhorace** [comment]
Even you dump it in a text file, I would be fine with that!

**Comment by lhorace** [comment]
JSON preferably

**Comment by tsujan** [comment]
> Even you dump it in a text file, I would be fine with that!

Adding such notifications to a test file is easy but it's only the first step. In fact, doing this job isn't difficult at all.

**Comment by tsujan** [comment]
The saving part is done:

![notifications](https://user-images.githubusercontent.com/981076/46902091-d6123300-cecb-11e8-9746-49e38ac18634.png)

The unattended notifications are saved in `~/.config/lxqt/notifications-list.conf` with the `QSettings` format (json should be possible too but I didn't find any LXQt component to use it -- maybe later).

Please tell me if you know of better/shorter labels/names!

The rest will be about adding a systray icon.

**Comment by Vladimir-csp** [comment]
Is there any reason to save notifications on storage? It's a dozen of kilobytes max.
If saving is necessary, then something like `${XDG_CACHE_HOME:-$HOME/.cache}` would be more appropriate location.

**Comment by tsujan** [comment]
> Is there any reason to save notifications on storage?

Oh, yes! Otherwise, how could unattended notifications of previous sessions be listed? They should be there until "recalled" or explicitly cleared. 

> It's a dozen of kilobytes max.

Yes, it should be light -- with 10 entries, it's 2.1 KiB here. For now, the max. number of entries can be up to 99. Also note that notification texts may be long and various notification actions may exist.

> If saving is necessary, then something like ${XDG_CACHE_HOME:-$HOME/.cache} would be more appropriate location.

I'm not sure.  I usually think about such details in the end -- if I don't forget them ;) We'll discuss that later.

**Comment by stefonarch** [comment]
2) context menu "show" and "clear all" ?
3) Number in overlay would be better
4) max 30? I see you can limit them in configuration after.

**Comment by tsujan** [comment]
> context menu "show" and "clear all" ?

"Clear all" is definitely needed. "Show" isn't needed because the notification is shown on clicking its item and then it's removed from the list (it'll be added again if still unattended and that may be a problem):

![tray](https://user-images.githubusercontent.com/981076/46905567-90bd2800-cf02-11e8-88ff-cc5b8799495e.png)

What should the text of each item be? Only human-readable date/time (ignore the ugly screenshot; it's only about testing) or also the notification title if any?

> Number in overlay would be better

Will think about it.

> max 30?

30 is much more acceptable than 90.

**Comment by Vladimir-csp** [comment]
1. Because it is technically cache. If you want it to be more permanent data, `${XDG_DATA_HOME:-$HOME/.local/share}`
2. list of notifications, 'Clear', 'Options'
3. I find the number of unread messages on Firetray Thunderbird plugin icon quite helpful.

Radical notion: is it possible to just display full notifications inside menu items and do away with extra step of opening them?

**Comment by tsujan** [comment]
> list of notifications, 'Clear', 'Options'

Oh, I'd forgotten about 'Options'. Thanks!

> Radical notion: is it possible to just display full notifications inside menu items and do away with extra step of opening them?

That radical notion had occurred to me too :) It's possible through a lot of work. The menu should contain a scroll-view and...

The code I have here isn't for general use yet -- I had some spare time today and wanted a "proof-of-concept" as soon as possible -- it still needs a lot of work. Your radical notion will come after that.

**Comment by lhorace** [comment]
What I need:

1. Source of notification [I can go to the source of the application that created the notification]
2. Summary of Text [If it's important enough that requires my attention]
3. Date/Time would be kewl but not very important to me. [That's added bonus but not really important to me] 

You guys, thanks, can decide how you want to show the notifications. The current notification system isn't great but I'm not complaining. I like to note doh, some source applications include 'activate', which brings the application to the foreground, open a new tab, and goes directly to the web page. Which is a very kewl feature

**Comment by lhorace** [comment]
> Is there any reason to save notifications on storage? It's a dozen of kilobytes max.

What if you had a sudden power failure? Or, you rebooted without acknowledging the notifications? Sometimes these notifications can be from family who in a emergency.

**Comment by tsujan** [comment]
@lhorace The list is saved; keeping it in memory is wrong, as you said.

After using it for hours, I found this really necessary: *an option for an exception list of apps*. It's VERY annoying that every unattended notification is added to the list.

**Comment by tsujan** [comment]
OK, it seems both usable and useful now. I'll use it for a day or two and then make a PR for testing and reviewing.

I implemented neither an item-view (@Vladimir-csp's suggestion) -- there's only a menu -- nor a number overlay -- just a tooltip. IMO, they could be made later if this feature is found handy and the menu doesn't seem to be sufficient (it does for me). The main idea is never missing a notification and it's implemented.

**Comment by tsujan** [comment]
The PR is here: https://github.com/lxqt/lxqt-notificationd/pull/103

**Comment by stefonarch** [comment]
Hm, compiled but I never see a new systray icon, are there other components to rebuild?

**Comment by tsujan** [comment]
@stefonarch First, thanks for testing!
As explained in the PR, the systray icon appears only if there's a missed notification. Make one with:
```
notify-send -a 'Downloader' -t 3000 -i emblem-downloads 'Download' '<b>Download Complete!</b>'
```
and leave it to disappear.

Of course, you should also restart `lxqt-notificationd` after installation.

**Comment by tsujan** [comment]
And don't forget to set a positive value for the "How many to save:" spin box. The default is zero.

**Comment by stefonarch** [comment]
Well the point was that I didn't activate it in the settings, default was 0...
Maybe set a default of  5 or 10?
For the rest works like a charm!
Clicking "Options" could point to "Advanced options" directly?
![schermata-16-19-38-35](https://user-images.githubusercontent.com/10681413/47035768-1b0fc100-d17b-11e8-830a-6c9dacc83953.png)

**Comment by tsujan** [comment]
> default was 0

Yes! That means this feature is disabled by default; hence its options among Advanced Settings.

> I think many people could miss it with default=0

Yes, unfortunately! I didn't want to enforce it as the default. Maybe that should be changed if it becomes popular and, in a sense, taken for granted, as it should be IMO.

> Clicking "Options" could point to "Advanced options" directly?

That needs a command-line option for `lxqt-config-notificationd`. Easy to add but I didn't want to go so far in this PR. It might be done in another PR, later.

**Comment by stefonarch** [comment]
Hm, for some reason excluding thunderbird doesn'work. It works for lxqt-panel, other apps I didn' test yet.

**Comment by tsujan** [comment]
@stefonarch It isn't always the binary name but what the notification announces. For Thunderbird it's `Thunderbird` (at least, with my version), for screengrab it's `Screen Grab`, etc. You should see what's written on the top of the notification. Don't add spaces before and after the name.

**Comment by stefonarch** [comment]
Yeah, I thought that and tried with uppercase and it works, so maybe a tooltip or an explanation in the settings about the name to use?

**Comment by tsujan** [comment]
Yes, a short tooltip is needed because the binary name can't be used there always (the same app may send notifications with different app names -- `notify-send` is a good example). I'll add it to the PR soon.

**Comment by stefonarch** [comment]
Hm, I have on custom script with a spacing in the name, and I see  no way to add it, with or without \
Same goes for Screen Grab.

**Comment by lhorace** [comment]
I created a PKGBUILD for those who might be interested in testing the new never_miss_notif patch for ArchLinux users.

```bash
_pkgname=lxqt-notificationd
pkgname=$_pkgname-never_miss_notif-git
pkgver=0.13.0.92.g951875e
pkgrel=1
pkgdesc="LXQt notification daemon and library. With never_miss_notif patch."
arch=("i686" "x86_64")
url="https://lxqt.org"
license=("LGPL2.1")
depends=("qt5-base" "liblxqt-git" "kwindowsystem")
makedepends=("git" "cmake" "qt5-tools" "lxqt-build-tools-git")
provides=("$_pkgname" "notification-daemon")
conflicts=("$_pkgname" "$_pkgname-git")
source=("git+https://github.com/lxqt/$_pkgname.git")
sha256sums=('SKIP')

prepare() {
	cd "$srcdir/$_pkgname"
	git checkout never_miss_notif
}

pkgver() {
	cd "$srcdir/$_pkgname"
	git describe --always | sed "s/-/./g"
}

build() {
	mkdir -p build
	cd build
	cmake "$srcdir/$_pkgname" \
		-DCMAKE_INSTALL_PREFIX=/usr
	make
}

package() {
	cd build
	make DESTDIR="$pkgdir" install
}
```

**Comment by lhorace** [comment]
For the past two days, for the most part, the new notification system has worked without any hiccups. Albeit, I'm not sure if this is by design or lack of functionality. When a notification get's hidden, stored, the `Activate` and `Open` buttons do nothing? This is just simply matter of convenience, for example, when I get E-Mail notification. `Open` would either launch or foreground my E-Mail client and directly linked to the _specific_ E-Mail message. This feature is really useful when you are subscribed to 12 or more mailing list, following specific discussions. Currently, often than not, the notification only show the thread subject, and I'm relegated to searching for the specific message the notification is referring too.

**Comment by tsujan** [comment]
> When a notification get's hidden, stored, the Activate and Open buttons do nothing?

That's because the remembered notification isn't related to or sent by a specific app anymore; it's just reproduced visually.

**Comment by lhorace** [comment]
> lack of functionality

Can it be implemented?

**Comment by tsujan** [comment]
I don't think so. Do you know of a notification daemon that does that? For any app?

**Comment by tsujan** [comment]
Moreover, the app may be closed or even removed after its notifications is saved.

**Comment by lhorace** [comment]
Gnome 3 on Ubuntu did it best and with time I'm sure I can find other DE environments that can do it too. On Ubuntu 18.04, I wouldn't acknowledge it on purpose and when I'm available (sometimes a day later) I click on it and it would foreground Firefox, open a new tab, and go to the page the notification was referring to.

**Comment by tsujan** [comment]
> Gnome 3 on Ubuntu did it best

With which apps? With their own email clients?

**Comment by lhorace** [comment]
On Ubuntu, at the time, the apps were: `Chromium` and `Firefox` with `push notifications`, and `Geary`

**Comment by tsujan** [comment]
Will investigate it later -- I'm very busy now -- but Geary's description: `A lightweight email client for the GNOME desktop`. I don't know about push notifications yet.

**Comment by tsujan** [comment]
@lhorace 

To know why it's impossible, install `dunstify`, save the following script, make its file executable, run it from a terminal, and click the notification button!

```sh
#!/bin/bash
myMsg=$(dunstify -A "blah blah,Click Here" -a 'A Script' -t 5000 -i folder 'Notification' "A useless notification.")
echo "I can do anything with $myMsg. Another script may send exactly the same notification and do something else :P"
exit
```

**Comment by lhorace** [comment]
I doubt it's impossible, but I would have to investigate what `dunstify` actually does first. Typically, not often the same message and it would be interesting to know what Gnome 3/Gnome based desktops like Solus and/or Deepin are doing under the hood.

**Comment by tsujan** [comment]
> I doubt it's impossible, but I would have to investigate what dunstify actually does first

This has nothing to do with dunstify. In the above example, I used it only to add a notification action *easily*.

The notification action (button) contains a message (a string), which is "emitted" when the action is invoked (the button is clicked). That click doesn't make a magic; an app should "listen" to it if it wants to do something based on it.

Clicking on hyperlinks inside labels has another story and can be implemented. However, it isn't related to this PR, which is only about not missing notifications.

**Comment by tsujan** [comment]
@agaida Shouldn't we set the max. unattended notification to a positive number (5, for example) by default? It's zero by default and, because of that, users might never find this feature, as @stefonarch [mentioned](https://github.com/lxqt/lxqt/issues/1573#issuecomment-430329334).

**Comment by agaida** [comment]
yes - i would suggest 10-20
